proposals/0000-pure-template-haskell.md

```diff
@@ -0,0 +1,294 @@
+And why should there be? It wouldn’t matter... except that it’s visible to the program being compiled.
+
+Pure Template Haskell would by no means deal with all the problems of cross-compiling with Template Haskell, but it would at least remove some of them. Cross-compilers could conceivably refuse to compile impure Template Haskell (although this would likely be too stringent in practice).
+
+### Recompilation avoidance
+
+Despite the fact that Template Haskell can perform IO, GHC currently uses a fairly optimistic recompilation check, which essentially assumes that the results of the IO are the same. 
+Users can use `addDependentFile` to declare a dependency on an external file, giving a limited improvement to recompilation checking.
+
+Pure Template Haskell would help in that the current recompilation check would actually be correct for modules that only have Pure Template Haskell. 
+We would still have the same difficulty for modules with impure Template Haskell, but since there would be far fewer, we could potentially use a more conservative recompilation check without having a terrible impact on the innocent majority.
+
+### Better motivation
+
+The above motivation is somewhat weak. 
+Alone, this proposal is more of a “cleanliness” proposal. However, the author believes it is an important step towards making TH easier for GHC to work with, which in the long run will be quite beneficial. 
+The motivation for this proposal is therefore missing some support from an (unwritten) larger manifesto on this topic.
+
+## Proposed Change Specification
+
+### 0. Terminology
+
+A Template Haskell quote is said to be "impure" if it has type `m a` where `m` has or requires a `MonadIO` instance.
+
+A Template Haskell splice is said to be "impure" if the code inside the splice has type `m a` where `m` has or requires a `MonadIO` instance.
+
+Note that today types for quotes and splice code are already inferred with appropriate constraints depending on what they use.
+
+### 1. Syntax
+
+A new syntax for top-level “impure splices” is introduced.
+
+- Impure untyped splice: `$!..`
+- Impure typed splice: `$$!..`
+
+We will elide the distinction between a splice being impure in the sense of "needing IO" and a splice being impure in the sense of "using impure splice syntax", so long as this is not confusing.
+
+Splices which do not use the impure splice syntax are called “normal” splices. This includes:
+- Splices of all kinds when `EnforceTemplateHaskellPurity` is disabled
+- Pure splices when `EnforceTemplateHaskellPurity` is enabled
+- Nested (non-top-level) splices
+
+### 2. Additions to `template-haskell`
+
+Introduce a new class `Reify`, containing the `reify` methods from `Quasi`:
+```
+–- could use MonadError here, but we’re going to impose a Reify m constraint at 
+-- the top level, and we probably want users to be able to pick their own 
+-- error monad for m if necessary
```

Comment:
I'm not sure this would work. Ultimately, users will write their splices in `PureQ` (I think), and `PureQ` can only offer the effects offered by `Reify`, so you have to add reporting here if you want users to be able to report errors.